### ISSUE TITLE
auto-refresh AAD token before the hour expiration

### ISSUE THREAD

**Issue by swanderz** [dbt-msft/dbt-sqlserver#156]
to use `authentication=cli`, you need to first log into to Azure using the Azure CLI using some form of `az login`. However, the token given by Azure Active Directory (AAD) expires after an hour (3600 seconds). So, if your long-running `dbt run` command has models starting to be created (i.e. new transaction) after 3600 seconds, `dbt run` will fail with one of the following messages:
- `Login failed for user ''. (18456) (SQLDriverConnect)")` (if using a service principal), or
- `token expired` (if using a user principal).

There are hacky workarounds, but the answer is to implement a token refresh. before the original token expires. @JCZuurmond any ideas for how this might work?

**Comment by JCZuurmond** [comment]
Always fun these kind of problems! 

Apparently you can set a [policy](https://docs.microsoft.com/en-us/azure/active-directory/develop/configure-token-lifetimes) with the token life time, but not having a powershell does not make that straightforward.

Part of the problem is that we [cache the token indefinitely](https://github.com/dbt-msft/dbt-sqlserver/blob/f2a480689825ac1194e26e6dc6427c4b6308ca02/dbt/adapters/sqlserver/connections.py#L200). We could make this "cache" time based.

However, if the cache would be time based, still the `open` method needs to be called more than once within a hour. I don't know how dbt handles long running queries. Does it simply wait for a response?

**Comment by JCZuurmond** [comment]
The execute is in [our control](https://github.com/dbt-msft/dbt-sqlserver/blob/f2a480689825ac1194e26e6dc6427c4b6308ca02/dbt/adapters/sqlserver/connections.py#L366) and dbt does not do anything with long running queries -  it simply calls the `execute` method.

So, if there is **one** query that runs for more than a hour, for example the creation of a single data model, then changing the caching method does not help. 

If multiple queries are ran, e.g. multiple models are created, each of them are created well within an hour, but all of them together take more than an hour, then improving the caching mechanism works.

**Comment by alieus** [comment]
In our case, It seems like we're faced with both scenarios: 

- multiple queries (data models) run, but not all queries are started within the 1 hour mark. So some models start running after the 1hr mark.
- Some of these queries for a single data model run for over an hour.

So @JCZuurmond, in this case, based on your points above, It looks like improving the caching method wouldn't help.

**Comment by swanderz** [comment]
@alieus  I think we can solve the first problem with token caching. The second isn't a really problem, right? As long as the transaction starts while the token is valid, it can go on for as long as it need AFAIK. The important part is to refresh the token when the expiry time to close before opening a new connection.

**Comment by JCZuurmond** [comment]
Yes, we do not know exactly what will happen with such a long running query. Still I expect it is likely that it will not work as the `cursor.execute` wait for response, so there is no chance to refresh the token.

For the multiple queries scenario replace [this line](https://github.com/dbt-msft/dbt-sqlserver/blob/f2a480689825ac1194e26e6dc6427c4b6308ca02/dbt/adapters/sqlserver/connections.py#L200) with something like:

``` python
if _TOKEN is None or (_TOKEN.expires_on - time.time()) < MAX_QUERY_TIME_IN_SECONDS:
    _TOKEN = azure_auth_function(credentials)
```

Then with `MAX_QUERY_TIME_IN_SECONDS` you can balance how often you refresh a token (caching) vs how long a query might take. I think setting it at 55 minutes (3300) makes sense. In any case, the value should be between an hour (3600) and 0 second.

**Comment by alieus** [comment]
@JCZuurmond  I experimented with commenting out [this line](https://github.com/dbt-msft/dbt-sqlserver/blob/5cebdaa9f8a40862db7769e43195e295743b13ba/dbt/adapters/sqlserver/connections.py#L283)  and that seems to solve the problem in both scenarios. Basically, don't even bother checking if we already have a token; for each new connection just create a new token. 

So your solution above definitely seems like a viable one.

**Comment by JCZuurmond** [comment]
Great. @swanderz , according to git you added the "caching" mechanism, see blame of [this line](https://github.com/dbt-msft/dbt-sqlserver/blob/f2a480689825ac1194e26e6dc6427c4b6308ca02/dbt/adapters/sqlserver/connections.py#L210). Is the caching necessary? Can it not simply be removed?

**Comment by swanderz** [comment]
resolved by #158
